File: src/backend/pcp/src/model/pcp_op_detalhe_model.py
from src import db
from src.model.produto_model import ProdutoModel


class PcpOpDetalheModel(db.Model):
    __tablename__ = 'pcp_op_detalhe'

    id = db.Column(db.Integer, primary_key=True)
    quantidade_produzir = db.Column(db.Float)
    quantidade_produzida = db.Column(db.Float)
    quantidade_entregue = db.Column(db.Float)
    custo_previsto = db.Column(db.Float)
    custo_realizado = db.Column(db.Float)
    id_pcp_op_cabecalho = db.Column(db.Integer, db.ForeignKey('pcp_op_cabecalho.id'))
    id_produto = db.Column(db.Integer, db.ForeignKey('produto.id'))

    produto_model = db.relationship('ProdutoModel', foreign_keys=[id_produto])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_pcp_op_cabecalho = data.get('idPcpOpCabecalho')
        self.id_produto = data.get('idProduto')
        self.quantidade_produzir = data.get('quantidadeProduzir')
        self.quantidade_produzida = data.get('quantidadeProduzida')
        self.quantidade_entregue = data.get('quantidadeEntregue')
        self.custo_previsto = data.get('custoPrevisto')
        self.custo_realizado = data.get('custoRealizado')

    def serialize(self):
        return {
            'id': self.id,
            'idPcpOpCabecalho': self.id_pcp_op_cabecalho,
            'idProduto': self.id_produto,
            'quantidadeProduzir': self.quantidade_produzir,
            'quantidadeProduzida': self.quantidade_produzida,
            'quantidadeEntregue': self.quantidade_entregue,
            'custoPrevisto': self.custo_previsto,
            'custoRealizado': self.custo_realizado,
            'produtoModel': self.produto_model.serialize() if self.produto_model else None,
        }

File: frontend/pcp/lib/app/data/model/pcp_op_detalhe_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:pcp/app/data/model/model_imports.dart';

class PcpOpDetalheModel {
	int? id;
	int? idPcpOpCabecalho;
	int? idProduto;
	double? quantidadeProduzir;
	double? quantidadeProduzida;
	double? quantidadeEntregue;
	double? custoPrevisto;
	double? custoRealizado;
	ProdutoModel? produtoModel;

	PcpOpDetalheModel({
		this.id,
		this.idPcpOpCabecalho,
		this.idProduto,
		this.quantidadeProduzir,
		this.quantidadeProduzida,
		this.quantidadeEntregue,
		this.custoPrevisto,
		this.custoRealizado,
		this.produtoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'quantidade_produzir',
		'quantidade_produzida',
		'quantidade_entregue',
		'custo_previsto',
		'custo_realizado',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Quantidade Produzir',
		'Quantidade Produzida',
		'Quantidade Entregue',
		'Custo Previsto',
		'Custo Realizado',
	];

	PcpOpDetalheModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idPcpOpCabecalho = jsonData['idPcpOpCabecalho'];
		idProduto = jsonData['idProduto'];
		quantidadeProduzir = jsonData['quantidadeProduzir']?.toDouble();
		quantidadeProduzida = jsonData['quantidadeProduzida']?.toDouble();
		quantidadeEntregue = jsonData['quantidadeEntregue']?.toDouble();
		custoPrevisto = jsonData['custoPrevisto']?.toDouble();
		custoRealizado = jsonData['custoRealizado']?.toDouble();
		produtoModel = jsonData['produtoModel'] == null ? ProdutoModel() : ProdutoModel.fromJson(jsonData['produtoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idPcpOpCabecalho'] = idPcpOpCabecalho != 0 ? idPcpOpCabecalho : null;
		jsonData['idProduto'] = idProduto != 0 ? idProduto : null;
		jsonData['quantidadeProduzir'] = quantidadeProduzir;
		jsonData['quantidadeProduzida'] = quantidadeProduzida;
		jsonData['quantidadeEntregue'] = quantidadeEntregue;
		jsonData['custoPrevisto'] = custoPrevisto;
		jsonData['custoRealizado'] = custoRealizado;
		jsonData['produtoModel'] = produtoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idPcpOpCabecalho = plutoRow.cells['idPcpOpCabecalho']?.value;
		idProduto = plutoRow.cells['idProduto']?.value;
		quantidadeProduzir = plutoRow.cells['quantidadeProduzir']?.value?.toDouble();
		quantidadeProduzida = plutoRow.cells['quantidadeProduzida']?.value?.toDouble();
		quantidadeEntregue = plutoRow.cells['quantidadeEntregue']?.value?.toDouble();
		custoPrevisto = plutoRow.cells['custoPrevisto']?.value?.toDouble();
		custoRealizado = plutoRow.cells['custoRealizado']?.value?.toDouble();
		produtoModel = ProdutoModel();
		produtoModel?.nome = plutoRow.cells['produtoModel']?.value;
	}	

	PcpOpDetalheModel clone() {
		return PcpOpDetalheModel(
			id: id,
			idPcpOpCabecalho: idPcpOpCabecalho,
			idProduto: idProduto,
			quantidadeProduzir: quantidadeProduzir,
			quantidadeProduzida: quantidadeProduzida,
			quantidadeEntregue: quantidadeEntregue,
			custoPrevisto: custoPrevisto,
			custoRealizado: custoRealizado,
		);			
	}

	
}